### PR TITLE
MDVRPTW instances

### PR DESCRIPTION
**Pull request by N-Wouda**
This PR adds the MDVRPTW instances of Vidal et al. (2013) - see the attached README for details. The instances have been converted to VRPLIB format, but the underlying data is unchanged.

Conversion script for future reference (might be handy sometime):

<details>

```python
import glob
from pathlib import Path
from typing import Any

import numpy as np
import vrplib


def parse(lines) -> dict[str, Any]:
    line = next(lines)
    _, veh_per_depot, num_cust, num_depots = map(int, line.split())
    dim = num_cust + num_depots

    max_durations = np.zeros(veh_per_depot * num_depots, dtype=int)
    capacities = np.zeros(veh_per_depot * num_depots, dtype=int)
    veh2depot = np.repeat(list((range(1, num_depots + 1))), veh_per_depot)

    for idx in range(num_depots):
        # All the same for each depot; these hold globally (I checked each 
        # instance to confirm this).
        line = next(lines)
        max_dur, veh_capacity = map(int, line.split())
        max_durations[veh_per_depot * idx: veh_per_depot * (idx + 1)] = max_dur
        capacities[veh_per_depot * idx: veh_per_depot * (idx + 1)] = veh_capacity

    assert len(np.unique(max_durations)) == 1
    assert len(np.unique(capacities)) == 1

    node_coords = np.zeros((dim, 2), dtype=int)
    time_windows = np.zeros((dim, 2), dtype=int)
    demands = np.zeros(dim, dtype=int)
    serv_durs = np.zeros(dim, dtype=int)

    for idx in range(num_depots, dim):  # first the customers..
        line = next(lines)
        _, x, y, serv_dur, demand, *_, tw_early, tw_late = map(float, line.split())
        node_coords[idx] = int(1000 * x), int(1000 * y)
        time_windows[idx] = tw_early, tw_late
        demands[idx] = demand
        serv_durs[idx] = serv_dur

    for idx in range(num_depots):  # ..and then the depots
        line = next(lines)
        _, x, y, serv_dur, demand, *_, tw_early, tw_late = map(float, line.split())
        node_coords[idx] = int(1000 * x), int(1000 * y)
        time_windows[idx] = tw_early, tw_late
        demands[idx] = demand
        serv_durs[idx] = serv_dur

    return {
        "DIMENSION": num_cust + num_depots,
        "VEHICLES": veh_per_depot * num_depots,
        "CAPACITY": np.unique(capacities).item(),
        "VEHICLES_MAX_DURATION": np.unique(max_durations).item(),
        "NODE_COORD_SECTION": node_coords,
        "DEMAND_SECTION": demands,
        "SERVICE_TIME_SECTION": serv_durs,
        "TIME_WINDOW_SECTION": time_windows,
        "VEHICLES_DEPOT_SECTION": veh2depot,
        "DEPOT_SECTION": list(range(1, num_depots + 1)),
    }


def main():
    for file in glob.iglob("pr*.txt"):
        file = Path(file)

        with open(file) as fh:
            parsed = {
                "NAME": file.stem.upper(),
                "COMMENT": "Based on Vidal et al. (2013)",
                "TYPE": "MDVRPTW",
                "EDGE_WEIGHT_TYPE": "EUC_2D",
            }
                        
            data = fh.readlines()
            parsed |= parse(iter(data))

            vrplib.write_instance(f"{file.stem.upper()}.vrp", parsed)


if __name__ == "__main__":
    main()
```

</details>

TODO:
- [x] Format instances VRPLIB style
- [ ] Hunt down some BKS

### CODE REVIEW

**Comment by N-Wouda** [comment]
@leonlan do you know if there are any BKS tracked for these instances? If so, where? If not, should we ask Thibaut for his BKS to kick start this?

**Comment by leonlan** [comment]
No, I don't know of any solutions unfortunately. This is a more common problem in general; I'm often only able to find the costs (see supplementary material [Vidal 2012](https://w1.cirrelt.ca/~vidalt/resources/UHGS_EC.pdf)) but not the actual solutions. Probably will have to ping Vidal for those.

**Comment by N-Wouda** [comment]
Yeah, I figured. That's a bit of a shame. I'll see what we can do about this once I continue working on MDVRPTW.

**Comment by N-Wouda** [comment]
I think we can finish https://github.com/PyVRP/PyVRP/pull/411 before this, but we'll need this as well to make sure that (a) the implementation in PyVRP 0.7.0 is completely correct, and (b) to benchmark PyVRP 0.7.0 in the first place.

**Comment by N-Wouda** [comment]
I've asked Thibaut for BKS on these instances.

**Comment by N-Wouda** [comment]
Useful: https://github.com/VROOM-Project/vroom-scripts/blob/master/src/mdvrp_to_json.py

**Comment by N-Wouda** [comment]
@leonlan could you have a look at these instances to see if you agree with the choice of field names? As it turns out, all vehicles have the same capacity/max durations, so I have added those as specifications, not sections.

**Comment by N-Wouda** [comment]
> I've asked Thibaut for BKS on these instances.

Haven't heard anything back BTW, so I'm not counting on this happening. Let's run PyVRP for a bit and then compare to the literature, I guess.

**Comment by N-Wouda** [comment]
I'll work on a PR for PyVRP to enable reading these instances.

File: MDVRPTW/PR11A.vrp [hunk 1]
```diff
@@ -0,0 +1,1515 @@
+NAME: PR11A
+COMMENT: Based on Vidal et al. (2013)
+TYPE: MDVRPTW
+EDGE_WEIGHT_TYPE: EUC_2D
+DIMENSION: 364
+VEHICLES: 40
+CAPACITY: 200
+VEHICLES_MAX_DURATION: 450
+NODE_COORD_SECTION
+1	2958	4357
```
Comment (leonlan) [inline MDVRPTW/PR11A.vrp:10]: The node coordinates are very large compared to the vehicles max duration/time windows. Is that correct?

**Review by leonlan — APPROVED** [review]
Just to summarize for myself what's new (correct me if wrong/missing things):
- `VEHICLES_MAX_DURATION` specification that specifies the maximum route duration for each vehicle. Can also be `VEHICLES_MAX_DURATION_SECTION` if the data is heterogeneous.
	- Edit: I recall that there is also a `DISTANCE` specification used by [CVRP instances](http://vrp.atd-lab.inf.puc-rio.br/media/com_vrp/instances/Li/Li_21.vrp) which is in the same spirit of maximum duration. Just an FYI, I prefer the new max duration because it's less ambiguous.
- `VEHICLES_DEPOT_SECTION` section. This specifies the start/end depot for each vehicle.
- `DEPOT_SECTION`: location indices of depots.

File: MDVRPTW/PR11A.vrp (hunk 1 above)
Comment (N-Wouda) [reply MDVRPTW/PR11A.vrp:10]: I had to round the X, Y coords from 3-decimal floats, so these are 1000x their original values. I should probably also round the other values to keep balance.

Comment (N-Wouda) [reply MDVRPTW/PR11A.vrp:10]: Thanks!

**Comment by N-Wouda** [comment]
> * `VEHICLES_MAX_DURATION` specification that specifies the maximum route duration for each vehicle. Can also be `VEHICLES_MAX_DURATION_SECTION` if the data is heterogeneous.
>   
>   * Edit: I recall that there is also a `DISTANCE` specification used by [CVRP instances](http://vrp.atd-lab.inf.puc-rio.br/media/com_vrp/instances/Li/Li_21.vrp) which is in the same spirit of maximum duration. Just an FYI, I prefer the new max duration because it's less ambiguous.

Yes, both the specification and section. For now our `read` method in PyVRP only understands the former, but that won't be hard to update if/when we actually have heterogeneous data.

> * `VEHICLES_DEPOT_SECTION` section. This specifies the start/end depot for each vehicle.

Si!

> * `DEPOT_SECTION`: location indices of depots.

This already existed, so not really new (CVRP and VRPTW also use this). Just has more than one value now.

**Comment by N-Wouda** [comment]
I merged this just yet and will run a try-out benchmark in a minute. Might change a few more things if they don't work out.

**Pull request merged**